src/coordinator/functions/logical/common.go

```diff
@@ -66,10 +67,55 @@ func hashFunc(on bool, names ...string) func(models.Tags) uint64 {
 	return func(tags models.Tags) uint64 { return tags.IDWithExcludes(names...) }
 }
 
-// Processor is implemented by each logical transform
-type Processor interface {
-	Process(lhs block.Block, rhs block.Block) (block.Block, error)
+const (
+	// AndType uses values from left hand side for which there is a value in right hand side with exactly matching label sets.
+	// Other elements are replaced by NaNs. The metric name and values are carried over from the left-hand side.
+	AndType = "and"
+
+	// OrType uses all values from left hand side, and appends values from the right hand side which do
+	// not have corresponding tags on the right
+	OrType = "or"
+
+	// UnlessType uses all values from lhs which do not exist in rhs
+	UnlessType = "unless"
+
+	initIndexSliceLength = 10
+)
+
+var (
+	errMismatchedBounds     = fmt.Errorf("block bounds are mismatched")
+	errMismatchedStepCounts = fmt.Errorf("block step counts are mismatched")
+	errConflictingTags      = fmt.Errorf("block tags conflict")
+)
+
+func combineMetadata(l, r block.Metadata) (block.Metadata, error) {
+	if !l.Bounds.Equals(r.Bounds) {
+		return block.Metadata{}, errMismatchedBounds
+	}
+
+	for k, v := range r.Tags {
+		if lVal, ok := l.Tags[k]; ok {
+			if lVal != v {
+				return block.Metadata{}, errConflictingTags
```

Comment:
what about tags which are in l but not in r ? Maybe add a comment to the function since its not clear what we're doing.

Reply:
Since it's returning `l` at the end, all tags which are in l will be present already; I'll add this comment